### PR TITLE
feat(smn): add new data source smn topics

### PR DESCRIPTION
**Pull request by Jason-Zhang9309**
**What this PR does / why we need it**:

add new data source smn topics

**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
add new data source smn topics
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST='./huaweicloud/services/acceptance/smn' TESTARGS='-run TestAccDataTopics_basic'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/smn -v -run TestAccDataTopics_basic -timeout 360m -parallel 4
=== RUN   TestAccDataTopics_basic
=== PAUSE TestAccDataTopics_basic
=== CONT  TestAccDataTopics_basic
--- PASS: TestAccDataTopics_basic (18.70s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/smn       18.800s
```

### CODE REVIEW

File: huaweicloud/services/smn/data_source_huaweicloud_smn_topic.go [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+	if len(filterTopics) < 1 {
+		return diag.Errorf("Your query returned no results. " +
+			"Please change your search criteria and try again.")
+	}
+
+	if len(filterTopics) > 1 {
+		return diag.Errorf("Your query returned more than one result." +
+			" Please try a more specific search criteria")
+	}
```
Comment (Lance52259) [inline huaweicloud/services/smn/data_source_huaweicloud_smn_topic.go]: 1. pay attention to the case (**your**).
2. total two sentences and other codes no more than 120 characters, can be written on a single line.

File: huaweicloud/services/smn/data_source_huaweicloud_smn_topic.go [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+				Type:     schema.TypeInt,
+				Computed: true,
+			},
+		},
+	}
+}
+
+func dataSourceTopicRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	config := meta.(*config.Config)
+	client, err := config.SmnV2Client(config.GetRegion(d))
```
Comment (Lance52259) [inline huaweicloud/services/smn/data_source_huaweicloud_smn_topic.go]: we can use a variable to storage region

File: huaweicloud/services/smn/data_source_huaweicloud_smn_topic.go [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+	}
+	tagClient.MoreHeaders = map[string]string{
+		"X-SMN-RESOURCEID-TYPE": "name",
+	}
+	if resourceTags, err := tags.Get(tagClient, "smn_topic", d.Get("name").(string)).Extract(); err == nil {
+		tagmap := utils.TagsToMap(resourceTags.Tags)
+		mErr = multierror.Append(mErr, d.Set("tags", tagmap))
+	} else {
+		log.Printf("[WARN] fetching tags of SMN topic failed: %s", err)
+	}
```
Comment (Lance52259) [inline huaweicloud/services/smn/data_source_huaweicloud_smn_topic.go]: how about this
```
resourceTags, err := tags.Get(tagClient, "smn_topic", topicGet.Name).Extract()
```
If we enter the name, we can use the response value to obtain tags.

File: docs/data-sources/smn_topic.md [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+* `region` - (Optional, String) Specifies the region in which to obtain the SMN topic. If omitted, the
+  provider-level region will be used.
+
+* `name` - (Optional, String) Specifies the name of the topic.
+
+* `topic_urn` - (Optional, String) Specifies the topic URN.
+
+* `display_name` - (Optional, String) Specifies the topic display name.
+
+* `enterprise_project_id` - (Optional, String) Specifies the enterprise project id of the SMN topic.
```
Comment (Lance52259) [inline docs/data-sources/smn_topic.md]: s/id/ID

File: docs/data-sources/smn_topic.md [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+## Attributes Reference
+
+In addition to all arguments above, the following attributes are exported:
+
+* `id` - The resource ID. The value is the topic URN.
+
+* `push_policy` - Message pushing policy. 0 indicates that the message sending fails and the message is cached in the
+  queue. 1 indicates that the failed message is discarded.
```
Comment (Lance52259) [inline docs/data-sources/smn_topic.md]: how about this
```
* `push_policy` - Message pushing policy.
  + **0**: indicates that the message sending fails and the message is cached in the queue.
  + **1**: indicates that the failed message is discarded.
```

File: huaweicloud/services/acceptance/smn/data_source_huaweicloud_smn_topic_test.go [hunk 6]
```diff
@@ -0,0 +1,55 @@
...
+	return fmt.Sprintf(`
+%s
+
+data "huaweicloud_smn_topic" "test" {
+  name = "%s"
+
+  depends_on = [
+    huaweicloud_smn_topic.topic_1
+  ]
+}
```
Comment (Lance52259) [inline huaweicloud/services/acceptance/smn/data_source_huaweicloud_smn_topic_test.go]: how about this
```
data "huaweicloud_smn_topic" "test" {
  name = huaweicloud_smn_topic.topic_1.name
}
```
this replaces the `depends_on` meta-argument.

File: huaweicloud/services/acceptance/smn/data_source_huaweicloud_smn_topic_test.go [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+import (
+	"fmt"
+	"testing"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance"
+)
+
+func TestAccDataSMNVTopic_basic(t *testing.T) {
```
Comment (Lance52259) [inline huaweicloud/services/acceptance/smn/data_source_huaweicloud_smn_topic_test.go]: The `SMN` is unnecessary if we define it in the packge name.

File: huaweicloud/services/acceptance/smn/data_source_huaweicloud_smn_topic_test.go [hunk 8]
```diff
@@ -0,0 +1,55 @@
+package huaweicloud
```
Comment (Lance52259) [inline huaweicloud/services/acceptance/smn/data_source_huaweicloud_smn_topic_test.go]: s/huaweicloud/smn

Comment (Jason-Zhang9309) [reply huaweicloud/services/acceptance/smn/data_source_huaweicloud_smn_topic_test.go]: fixed

File: huaweicloud/services/acceptance/smn/data_source_huaweicloud_smn_topic_test.go (hunk 7 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/acceptance/smn/data_source_huaweicloud_smn_topic_test.go]: sure

File: huaweicloud/services/acceptance/smn/data_source_huaweicloud_smn_topic_test.go (hunk 6 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/acceptance/smn/data_source_huaweicloud_smn_topic_test.go]: the value of `huaweicloud_smn_topic.topic_1.name` is already known before apply, which means this equals to using the name directly and can not avoid error. Using depends_on is fine. Don't pay too much attention to it.

File: docs/data-sources/smn_topic.md (hunk 5 above)
Comment (Jason-Zhang9309) [reply docs/data-sources/smn_topic.md]: sure

File: docs/data-sources/smn_topic.md (hunk 4 above)
Comment (Jason-Zhang9309) [reply docs/data-sources/smn_topic.md]: fixed

File: huaweicloud/services/smn/data_source_huaweicloud_smn_topic.go (hunk 3 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/smn/data_source_huaweicloud_smn_topic.go]: fixed

File: huaweicloud/services/smn/data_source_huaweicloud_smn_topic.go (hunk 1 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/smn/data_source_huaweicloud_smn_topic.go]: deleted

File: huaweicloud/services/smn/data_source_huaweicloud_smn_topic.go (hunk 2 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/smn/data_source_huaweicloud_smn_topic.go]: ok

File: huaweicloud/provider.go [hunk 9]
```diff
@@ -445,6 +445,8 @@ func Provider() *schema.Provider {
 
 			"huaweicloud_servicestage_component_runtimes": servicestage.DataSourceComponentRuntimes(),
 
+			"huaweicloud_smn_topic": smn.DataSourceTopic(),
```
Comment (ShiChangkuo) [inline huaweicloud/provider.go]: all names of the data source, including functions, files and docs, should be updated to **topics**

File: huaweicloud/services/smn/data_source_huaweicloud_smn_topic.go [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+	if err != nil {
+		return diag.Errorf("filter topics failed: %s", err)
+	}
+
+	ids := make([]string, len(filterTopics))
+	stateTopics := make([]map[string]interface{}, len(filterTopics))
+	// create tagClient to fetch tags
+	tagClient, err := config.SmnV2TagClient(region)
+	if err != nil {
+		return diag.Errorf("error creating SMN tag client: %s", err)
```
Comment (niuzhenguo) [inline huaweicloud/services/smn/data_source_huaweicloud_smn_topic.go]: should we consider to suppress this error as tags is an extra attribute of the topic?

Comment (Jason-Zhang9309) [reply huaweicloud/services/smn/data_source_huaweicloud_smn_topic.go]: getting tags error is suppressed, no need to suppress the error of creating client

File: huaweicloud/provider.go (hunk 9 above)
Comment (Jason-Zhang9309) [reply huaweicloud/provider.go]: fixed

File: huaweicloud/services/smn/data_source_huaweicloud_smn_topics.go [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+
+	filterTopics, err := utils.FilterSliceWithField(allTopics, filter)
+	if err != nil {
+		return diag.Errorf("filter topics failed: %s", err)
+	}
+
+	ids := make([]string, len(filterTopics))
+	stateTopics := make([]map[string]interface{}, len(filterTopics))
+	// create tagClient to fetch tags
+	tagClient, err := config.SmnV2TagClient(region)
```
Comment (ShiChangkuo) [inline huaweicloud/services/smn/data_source_huaweicloud_smn_topics.go]: it's better to put tagClient together with smnClient.

File: huaweicloud/services/smn/data_source_huaweicloud_smn_topics.go [hunk 12]
```diff
@@ -0,0 +1,160 @@
...
+	if err != nil {
+		return diag.Errorf("error creating SMN tag client: %s", err)
+	}
+	tagClient.MoreHeaders = map[string]string{
```
Comment (ShiChangkuo) [inline huaweicloud/services/smn/data_source_huaweicloud_smn_topics.go]: why we need the header? I can not find the header in docs

File: huaweicloud/services/smn/data_source_huaweicloud_smn_topics.go (hunk 11 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/smn/data_source_huaweicloud_smn_topics.go]: sure

File: huaweicloud/services/smn/data_source_huaweicloud_smn_topics.go (hunk 12 above)
Comment (Jason-Zhang9309) [reply huaweicloud/services/smn/data_source_huaweicloud_smn_topics.go]: this is needed, see it here https://support.huaweicloud.com/intl/en-us/api-smn/BatchCreateOrDeleteResourceTags.html

Comment (ShiChangkuo) [reply huaweicloud/services/smn/data_source_huaweicloud_smn_topics.go]: got it

**Pull request merged**